### PR TITLE
feat: Use the trainee's TIS ID from token

### PR DESCRIPTION
**Pull request by Judge40**
Currently the trainee ID is provided as part of the request, but this
would potentially allow users to access data that is not theirs. Instead
the authentication token should be used to retrieve the logged in user's
TIS ID directly.

TISNEW-3833

### CODE REVIEW

**Comment by Judge40** [comment]
Created as draft due to API Gateway changes that are required for this to be merged successfully.

**Review by uzairbaig — APPROVED** [review]

File: src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Map;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+
+@Slf4j
+public class AuthTokenUtil {
```
Comment (paulhoang-hee) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java:35]: My preference would be to turn this into a spring bean and remove the static keywords

File: src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  private static final String TIS_ID_ATTRIBUTE = "custom:tisId";
+
+  private static final ObjectMapper mapper = new ObjectMapper();
+
+  private AuthTokenUtil() {
```
Comment (paulhoang-hee) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java:41]: if we're dealing with tokens, should we not validate them against the secret?

File: src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java (hunk 1 above)
Comment (Judge40) [reply src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java:35]: You're probably right, old habits die hard.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java (hunk 2 above)
Comment (Judge40) [reply src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java:41]: Wasn't sure about this when I was doing it to be honest. The expectation is that the service itself will always be private and public access is only via API Gateway. Given the API Gateway will authorize the token I've assumed, perhaps naïvely, that the token can be trusted within this service to keep things fairly simple.

It does rely on the service being kept private and only ever accessible through the API Gateway, I suppose there is room for error there.

**Review by DorisWongNHS — APPROVED** [review]

File: src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java (hunk 2 above)
Comment (paulhoang-hee) [reply src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java:41]: if the api gateway does the jwt validation then I don't see any issue with this. last thing you want is to duplicate this validation code everywhere in other services - yay for shared modules

File: src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Verify that the token contains a valid trainee TIS ID and that it matches the expected ID.
+   *
+   * @param requestedTraineeTisId The ID to validate against.
+   * @param token                 The token to use.
+   * @param <T>                   The type of the response body.
+   * @return A ResponseEntity with a suitable error status, or empty if valid.
+   */
+  public static <T> Optional<ResponseEntity<T>> verifyTraineeTisId(String requestedTraineeTisId,
```
Comment (paulhoang-hee) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/api/util/AuthTokenUtil.java:69]: I'd be tempted to remove the HTTP related stuff here and just return a boolean here to state if its valid

**Review by paulhoang-hee — APPROVED** [review]
good to go

**Pull request merged**